modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/CurrencySnapshotCreator.scala

```diff
@@ -50,137 +50,128 @@ trait CurrencySnapshotCreator[F[_]] {
-
-        maybeLastDataApplication = lastArtifact.data
-
-        maybeLastDataState <- (maybeLastDataApplication, maybeDataApplicationService).mapN {
-          case (lastDataApplication, (_, service)) =>
-            service
-              .deserializeState(lastDataApplication)
-              .flatTap {
-                case Left(err) => logger.warn(err)(s"Cannot deserialize custom state")
-                case Right(a)  => logger.info(s"Deserialized state: $a")
-              }
-              .map(_.toOption)
-              .handleErrorWith(err =>
-                logger.error(err)(s"Unhandled exception during deserialization data application, fallback to empty state") >>
-                  none[DataState].pure[F]
-              )
-        }.flatSequence
-
-        maybeNewDataState <- (maybeDataApplicationService, maybeLastDataState).mapN {
-          case ((nodeContext, service), lastState) =>
-            implicit val context: L0NodeContext[F] = nodeContext
-            NonEmptyList
-              .fromList(dataBlocks.distinctBy(_.value.roundId))
-              .map(_.flatMap(_.value.updates))
-              .map { updates =>
-                service
-                  .validateData(lastState, updates)
-                  .flatTap { validated =>
-                    Applicative[F].unlessA(validated.isValid)(logger.warn(s"Data application is invalid, errors: ${validated.toString}"))
-                  }
-                  .map(_.isValid)
-                  .handleErrorWith(logger.error(_)(s"Unhandled exception during validating data application, assumed as invalid").as(false))
-                  .ifF(updates.toList, List.empty[Signed[DataUpdate]])
+      rewards: Option[Rewards[F, CurrencySnapshotStateProof, CurrencyIncrementalSnapshot, CurrencySnapshotEvent]],
+      facilitators: Set[PeerId]
+    ): F[CurrencySnapshotCreationResult[CurrencySnapshotEvent]] = {
+
+      val maxArtifactSize = maxProposalSizeInBytes(facilitators)
+
+      def createProposalWithSizeLimit(
+        eventsForAcceptance: Set[CurrencySnapshotEvent],
+        sizeRejectedEvents: Set[CurrencySnapshotEvent] = Set.empty[CurrencySnapshotEvent]
+      ): F[CurrencySnapshotCreationResult[CurrencySnapshotEvent]] =
+        for {
+          lastArtifactHash <- lastArtifact.value.hashF
+          currentOrdinal = lastArtifact.ordinal.next
+
+          currentEpochProgress = trigger match {
+            case EventTrigger => lastArtifact.epochProgress
+            case TimeTrigger  => lastArtifact.epochProgress.next
+          }
+          lastActiveTips <- lastArtifact.activeTips
+          lastDeprecatedTips = lastArtifact.tips.deprecated
+          maybeLastDataApplication = lastArtifact.data
+
+          (blocks: List[Signed[Block]], dataBlocks: List[Signed[DataApplicationBlock]]) =
+            eventsForAcceptance
+              .filter(_.isLeft || dataApplicationSnapshotAcceptanceManager.isDefined)
+              .toList
+              .partitionMap(identity)
```

Comment:
From what I remember @marcinwadon decided to go for `isDefined` because it fits the concept: We should keep `DataApplicationBlock`s for acceptance only when `dataApplicationSnapshotAcceptanceManager` is **defined** by a metagraph developer. Otherwise, we do not know how to process such blocks.

I am happy with both ways of writing that part so maybe let's ask the others.

Reply:
@sdavidp imo this filter looks clear. we want to continue iif the data application is defined, so the implementation for metagraph is provided

Reply:
@marcinwadon Then maybe an `if .isDefined` instead of `match`? It looks confusing when the filter is checking if it's defined for every element of the list when we only need to check it once.

Reply:
>checking if it's defined for every element of the list when we only need to check it once.

You convinced me with this :+1: